08-rails/testing-db-in-controllers.md

```diff
@@ -4,15 +4,52 @@
 
 By the end of this lesson you should be able to:
 
-- write tests to verify controller actions make the proper changes to the database under nominal conditions
-- write tests to verify that controller actions respond properly to invalid or incorrect input
+- Describe and use the `must_change` Minitest matcher
+- Understand how to test nominal and edge cases for database changes
 
-## Content
+## The Data, It's a Changin'
 
-It's important to ensure that controller actions appropriately change the related model for the `destroy`, `create` and `update` actions.  To do so we will use the rails `must_change` matcher and see how to use the it.
+When we run a `POST` action, how do we know that the database is getting updated? Even if we know it's getting updated, how can we check that the data within is correct?
 
-`must_change` is called on a block, just like `must_raise`.  We give it a string which it evaluates as Ruby code before and after the block executes.  The second argument is the amount by which the 1st argument changes when the block executes.  
+One of the mandates of TDD is making sure that all the code that we've written is trustworthy, so that we can pass it off to other programmers with confidence. To do so we will use the rails `must_change` matcher to take a peek at our database during tests.
 
+Like its cousin `must_raise`, `must_change` is called on a block. We hand it two arguments: a string, which is evaluated as ruby code, and a number, which indicates how much we expect the value to be altered by. A single call to `must_change` might look like this:
+
+```ruby
+volume = 6
+expect {
+  volume += 5
+}.must_change 'volume', 5
+```
+
+This might seem pretty obvious, as we trust addition in ruby to do what we ask of it, but we can leverage it to check different, more complicated types of operations. Let's pretend that our current project could check out a book using a route with the following route:
+
+```ruby
+#routes.rb
+
+patch '/books/:id/check_out', to: 'books#check_out', as: 'checkout'
+```
+
+We'll also pretend, for the sake of this example, that we have a field in our `Book` called `in_stock` that we represent as a integer. When we check out a `Book` using this path, we expect that `somebook.in_stock` is reduced by one. The test might look something like this:
+
+```ruby
+describe "book_checkouts" do
+  it "can correctly check out a book" do
+    book = Book.first
+
+    expect{
+      patch checkout_path, params: book.id
+    }.must_change 'book.in_stock', 1
+  end
+end
+```
+
+A couple of things worth noting: 
+
+- `must_change` only works with numeric values.
+- There is no difference to minitest between something going up and something going down in value. In other words, it checks only the absolute value of the change.
```

Comment:
`must_change` does care if it goes up or down.

Reply:
So!  In `minitest-rails` 3.x `must_change` cares about sign.  In `minitest-rails` 5.x (they skipped 4 to line up with Rails versions) `must_change` now does the Rails `assert_changes` which is based on absolute value.

The old assertion is now `must_differ` and continues to do the useful thing: https://blowmage.com/2019/06/18/minitest-rails-releases  (Minitest 5.x was released in June of this year.)